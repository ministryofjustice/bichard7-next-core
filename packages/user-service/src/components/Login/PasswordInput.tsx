interface Props {
  label?: string
  labelSize?: "s" | "m" | "l"
  name: string
  hint?: string
  className?: string
  error?: string | boolean
  width?: "5" | "10" | "20" | "30"
  id?: string
  readOnly?: boolean
  disabled?: boolean
  mandatory?: boolean
  value?: string
  optionalProps?: object
}

const PasswordInput = ({
  label,
  labelSize,
  hint,
  error,
  className,
  name,
  width,
  id,
  readOnly = false,
  disabled = false,
  value,
  mandatory = false,
  optionalProps = {}
}: Props) => {
  const hintElementId = `${name}-hint`
  const errorElementId = `${name}-error`
  const inputId = id || name

  const hasError = error && !readOnly

  const formGroupClasses = [
    "govuk-form-group",
    "govuk-password-input",
    className,
    hasError ? "govuk-form-group--error" : null
  ]
    .filter(Boolean)
    .join(" ")

  const widthClass = width ? `govuk-input--width-${width}` : null

  const inputClasses = [
    "govuk-input",
    "govuk-password-input__input",
    "govuk-js-password-input-input",
    widthClass,
    hasError ? "govuk-input--error" : null
  ]
    .filter(Boolean)
    .join(" ")

  const labelClassName = labelSize ? `govuk-label--${labelSize}` : null

  const labelClasses = ["govuk-label", labelClassName].filter(Boolean).join(" ")

  const ariaDescribedBy = [hint ? hintElementId : null, hasError ? errorElementId : null].filter(Boolean).join(" ")
  return (
    <div className={formGroupClasses} data-module="govuk-password-input">
      {label && (
        <label className={labelClasses} htmlFor={name}>
          {label}
          {mandatory && " *"}
        </label>
      )}

      {hint && (
        <div id={hintElementId} className="govuk-hint">
          {hint}
        </div>
      )}
      {!!hasError && typeof error === "string" && (
        <span data-test={`password-input_${name}-error`} id={errorElementId} className="govuk-error-message">
          <span className="govuk-visually-hidden">{"Error:"}</span> {error}
        </span>
      )}

      <div className="govuk-input__wrapper govuk-password-input__wrapper">
        <input
          className={inputClasses}
          id={inputId}
          name={name}
          data-test={`password-input_${name}`}
          type="password"
          defaultValue={value}
          readOnly={readOnly}
          disabled={disabled}
          autoComplete="current-password"
          autoCapitalize="none"
          spellCheck="false"
          aria-describedby={ariaDescribedBy}
          {...optionalProps}
        />

        <button
          className="govuk-button govuk-button--secondary govuk-password-input__toggle govuk-js-password-input-toggle"
          type="button"
          data-module="govuk-button"
          aria-controls={inputId}
          aria-label="Show password"
        >
          {"Show"}
        </button>
      </div>
    </div>
  )
}

export default PasswordInput
