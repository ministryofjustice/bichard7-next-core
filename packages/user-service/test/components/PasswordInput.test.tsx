import { render } from "@testing-library/react"
import PasswordInput from "components/Login/PasswordInput"

describe("PasswordInput", () => {
  it("should render correctly with required props (label and name)", () => {
    const { container } = render(<PasswordInput label="Test Label" name="testPassword" />)
    expect(container).toMatchSnapshot()
  })

  it("should render with error styling and an error message when 'error' is a string", () => {
    const { container } = render(
      <PasswordInput label="Password" name="testPassword" error="You must enter a stronger password" />
    )
    expect(container).toMatchSnapshot()
  })

  it("should render with a hint, a specific width class, and no error", () => {
    const { container } = render(<PasswordInput label="Secret Code" name="secretCode" hint="Keep it safe" width="10" />)
    expect(container).toMatchSnapshot()
  })

  it("should render with a large label size class", () => {
    const { container } = render(<PasswordInput label="Large Label" name="largeLabelPass" labelSize="l" />)
    expect(container).toMatchSnapshot()
  })
})
