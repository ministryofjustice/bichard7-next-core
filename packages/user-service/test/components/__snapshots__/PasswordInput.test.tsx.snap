// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PasswordInput should render correctly with required props (label and name) 1`] = `
<div>
  <div
    class="govuk-form-group govuk-password-input"
    data-module="govuk-password-input"
  >
    <label
      class="govuk-label"
      for="testPassword"
    >
      Test Label
    </label>
    <div
      class="govuk-input__wrapper govuk-password-input__wrapper"
    >
      <input
        aria-describedby=""
        autocapitalize="none"
        autocomplete="current-password"
        class="govuk-input govuk-password-input__input govuk-js-password-input-input"
        data-test="password-input_testPassword"
        id="testPassword"
        name="testPassword"
        spellcheck="false"
        type="password"
      />
      <button
        aria-controls="testPassword"
        aria-label="Show password"
        class="govuk-button govuk-button--secondary govuk-password-input__toggle govuk-js-password-input-toggle"
        data-module="govuk-button"
        type="button"
      >
        Show
      </button>
    </div>
  </div>
</div>
`;

exports[`PasswordInput should render with a hint, a specific width class, and no error 1`] = `
<div>
  <div
    class="govuk-form-group govuk-password-input"
    data-module="govuk-password-input"
  >
    <label
      class="govuk-label"
      for="secretCode"
    >
      Secret Code
    </label>
    <div
      class="govuk-hint"
      id="secretCode-hint"
    >
      Keep it safe
    </div>
    <div
      class="govuk-input__wrapper govuk-password-input__wrapper"
    >
      <input
        aria-describedby="secretCode-hint"
        autocapitalize="none"
        autocomplete="current-password"
        class="govuk-input govuk-password-input__input govuk-js-password-input-input govuk-input--width-10"
        data-test="password-input_secretCode"
        id="secretCode"
        name="secretCode"
        spellcheck="false"
        type="password"
      />
      <button
        aria-controls="secretCode"
        aria-label="Show password"
        class="govuk-button govuk-button--secondary govuk-password-input__toggle govuk-js-password-input-toggle"
        data-module="govuk-button"
        type="button"
      >
        Show
      </button>
    </div>
  </div>
</div>
`;

exports[`PasswordInput should render with a large label size class 1`] = `
<div>
  <div
    class="govuk-form-group govuk-password-input"
    data-module="govuk-password-input"
  >
    <label
      class="govuk-label govuk-label--l"
      for="largeLabelPass"
    >
      Large Label
    </label>
    <div
      class="govuk-input__wrapper govuk-password-input__wrapper"
    >
      <input
        aria-describedby=""
        autocapitalize="none"
        autocomplete="current-password"
        class="govuk-input govuk-password-input__input govuk-js-password-input-input"
        data-test="password-input_largeLabelPass"
        id="largeLabelPass"
        name="largeLabelPass"
        spellcheck="false"
        type="password"
      />
      <button
        aria-controls="largeLabelPass"
        aria-label="Show password"
        class="govuk-button govuk-button--secondary govuk-password-input__toggle govuk-js-password-input-toggle"
        data-module="govuk-button"
        type="button"
      >
        Show
      </button>
    </div>
  </div>
</div>
`;

exports[`PasswordInput should render with error styling and an error message when 'error' is a string 1`] = `
<div>
  <div
    class="govuk-form-group govuk-password-input govuk-form-group--error"
    data-module="govuk-password-input"
  >
    <label
      class="govuk-label"
      for="testPassword"
    >
      Password
    </label>
    <span
      class="govuk-error-message"
      data-test="password-input_testPassword-error"
      id="testPassword-error"
    >
      <span
        class="govuk-visually-hidden"
      >
        Error:
      </span>
       
      You must enter a stronger password
    </span>
    <div
      class="govuk-input__wrapper govuk-password-input__wrapper"
    >
      <input
        aria-describedby="testPassword-error"
        autocapitalize="none"
        autocomplete="current-password"
        class="govuk-input govuk-password-input__input govuk-js-password-input-input govuk-input--error"
        data-test="password-input_testPassword"
        id="testPassword"
        name="testPassword"
        spellcheck="false"
        type="password"
      />
      <button
        aria-controls="testPassword"
        aria-label="Show password"
        class="govuk-button govuk-button--secondary govuk-password-input__toggle govuk-js-password-input-toggle"
        data-module="govuk-button"
        type="button"
      >
        Show
      </button>
    </div>
  </div>
</div>
`;
